modules/stochastic_tools/src/surrogates/PolynomialChaos.C

```diff
@@ -20,46 +20,79 @@ PolynomialChaos::validParams()
-    _quad_sampler(nullptr),
-    _tuple(generateTuple(getParam<std::vector<DistributionName>>("distributions").size(), _order)),
-    _ncoeff(_tuple.size()),
-    _coeff(_ncoeff)
+    _ndim(declareModelData<unsigned int>("_ndim")),
+    _tuple(declareModelData<std::vector<std::vector<unsigned int>>>("_tuple")),
+    _ncoeff(declareModelData<std::size_t>("_ncoeff")),
+    _coeff(declareModelData<std::vector<Real>>("_coeff")),
+    _poly(declareModelData<std::vector<std::unique_ptr<const PolynomialQuadrature::Polynomial>>>(
+        "_poly"))
 {
-  MooseUtils::linearPartitionItems(
-      _ncoeff, n_processors(), processor_id(), _n_local_coeff, _st_local_coeff, _end_local_coeff);
 }
 
 void
 PolynomialChaos::initialSetup()
 {
-  SurrogateModel::initialSetup();
-
-  // Special circumstances if sampler is quadrature
-  _quad_sampler = dynamic_cast<QuadratureSampler *>(_sampler);
+  if (isTraining())
+  {
+    // Setup data needed for training
+    _order = getParam<unsigned int>("order");
+    _tuple = generateTuple(getParam<std::vector<DistributionName>>("distributions").size(), _order);
+    _ncoeff = _tuple.size();
+    _coeff.resize(_ncoeff, 0);
+    _values_ptr =
+        &getVectorPostprocessorValue("results_vpp", getParam<std::string>("results_vector"));
+
+    // Sampler
+    _sampler = &getSamplerByName(getParam<SamplerName>("training_sampler"));
+    _ndim = _sampler->getNumberOfCols();
+
+    // Special circumstances if sampler is quadrature
+    _quad_sampler = dynamic_cast<QuadratureSampler *>(_sampler);
+
+    // Check if sampler dimensionality matches number of distributions
+    std::vector<DistributionName> dname = getParam<std::vector<DistributionName>>("distributions");
+    if (dname.size() != _ndim)
+      mooseError("Sampler number of columns does not match number of inputted distributions.");
+    for (auto nm : dname)
+      _poly.push_back(PolynomialQuadrature::makePolynomial(&getDistributionByName(nm)));
+  }
 
-  std::vector<DistributionName> dname = getParam<std::vector<DistributionName>>("distributions");
-  // Check if sampler dimensionality matches number of distributions
-  if (dname.size() != _ndim)
-    mooseError("Sampler number of columns does not match number of inputted distributions.");
-  for (auto nm : dname)
-    _poly.push_back(PolynomialQuadrature::makePolynomial(&getDistributionByName(nm)));
+  // Setup parallel partitioning of coefficients
+  MooseUtils::linearPartitionItems(_ncoeff,
+                                   n_processors(),
+                                   processor_id(),
+                                   _n_local_coeff,
+                                   _local_coeff_begin,
+                                   _local_coeff_end);
+
+  // If training data is loaded, it is store in a single file so the data must be partitioned
+  if (!isTraining() && n_processors() > 1)
+    _coeff.assign(_coeff.begin() + _local_coeff_begin, _coeff.begin() + _local_coeff_end);
```

Comment:
No, I just dropped this in at the last minute. I need to come up with an API for outputting as well. I think if you were to run in parallel now then only the part from rank 0 would show up in the file.

I think we will need to add "parallel_type" to SurrogateModel objects as well. And the output will need to be aware and do the correct thing. The SurrogateModel will need to be able to run in parallel and write out a single file or one file per process. Also, when loading it should be able to read a single file and setup a distributed model or read multiple files into a distributed run. We also need to thing about training on 100 cores, but evaluating on 10.

Initially, I want to just focus on a single file load/store, but handle the serialization and de-serialization of the model data. For this the SurrogateModel will probably need an "onLoad" and "onStore" methods to handle the data. I am still thinking about the API for this...it get complicated because if you put all the data on rank 0 for output but keep going in the solve/training now the data is not distributed, etc.

I probably need to come up with some test cases so I can understand what needs to be supported. Perhaps we pick the most common case and error on the others first.

Reply:
Wow, parallel stuff is hard. Would it help if I came up with a way to distribute the coefficients? I think this would require communication on each call to `evaluate`. Perhaps we could add `evaluateInitialize` and `evaluateFinalize` to take care of this. I have a feeling this will be useful for other surrogate models.

Reply:
Like I said on the call. I thought it was distributed, so this change is just wrong. For now lets assume that all training data is "replicated". This will make things a lot easier as we ramp up.

I think I will add "parallel_type" put only have "replicated" as an option. I will also add a parallel test where it is trained with 3 cores and executed with 2.